### PR TITLE
Selection: add shallowSelect and shallowSelectAll

### PR DESCRIPTION
**Pull request by c-frater**
#104

### CODE REVIEW

File: modules/selection/main/index.coffee [hunk 1]
```diff
@@ -16,9 +16,41 @@ getMethod = (node, methodName) ->
...
 
+getMatches = (node) ->
+  # XXX: use hx.vendor
+  matchesMethod =
+    if node.matches then 'matches'
+    else if node.matchesSelector then 'matchesSelector'
+    else if node.mozMatchesSelector then 'mozMatchesSelector'
+    else if node.webkitMatchesSelector then 'webkitMatchesSelector'
+    else if node.oMatchesSelector then 'oMatchesSelector'
+    else if node.msMatchesSelector then 'msMatchesSelector'
```
Comment (georgesimms) [inline modules/selection/main/index.coffee]: can we use hx.vendor here ?

File: modules/selection/main/index.coffee [hunk 2]
```diff
@@ -16,9 +16,41 @@ getMethod = (node, methodName) ->
...
+
+  if matchesMethod then getMethod(node, matchesMethod)
+  else (selector) ->
+    matchingNodes = getMethod(this, 'querySelector').call(this, selector)
+    i = matchingNodes.length
+    i while --i >= 0 and matchingNodes.item(i) isnt this
+    i > -1
```
Comment (georgesimms) [inline modules/selection/main/index.coffee]: The indentation of this if block is a bit hard to follow, maybe

```
if matchesMethod
  getMethod node, matchesMethod
else
  (selector) ->
    #stuff

File: modules/selection/main/index.coffee [hunk 3]
```diff
@@ -16,9 +16,41 @@ getMethod = (node, methodName) ->
...
+    else if node.mozMatchesSelector then 'mozMatchesSelector'
+    else if node.webkitMatchesSelector then 'webkitMatchesSelector'
+    else if node.oMatchesSelector then 'oMatchesSelector'
+    else if node.msMatchesSelector then 'msMatchesSelector'
+
+  if matchesMethod then getMethod(node, matchesMethod)
+  else (selector) ->
+    matchingNodes = getMethod(this, 'querySelector').call(this, selector)
+    i = matchingNodes.length
+    i while --i >= 0 and matchingNodes.item(i) isnt this
```
Comment (georgesimms) [inline modules/selection/main/index.coffee]: * What is the type of matchingNodes here ? is it a pseudo-array ?
* Why do we start at the end and work backward ?
* Can we
```
this in Array::slice.call matchingNodes
```
?

File: modules/selection/main/index.coffee [hunk 4]
```diff
@@ -16,9 +16,41 @@ getMethod = (node, methodName) ->
...
+  for child in node.children
+    if matchFn.call(child, selector)
+      return child
+
+shallowSelectAll = (selector, node) ->
+  matchFn = getMatches(node)
+  matchingNodes = []
+  for child in node.children
+    if matchFn.call(child, selector)
+      matchingNodes.push child
```
Comment (georgesimms) [inline modules/selection/main/index.coffee]: Again, maybe
```
Array::slice.call node.children
  .filter (selector) -> matchFn.call child, selector
```

File: modules/selection/main/index.coffee [hunk 5]
```diff
@@ -16,9 +16,41 @@ getMethod = (node, methodName) ->
...
+    if node.matches then 'matches'
+    else if node.matchesSelector then 'matchesSelector'
+    else if node.mozMatchesSelector then 'mozMatchesSelector'
+    else if node.webkitMatchesSelector then 'webkitMatchesSelector'
+    else if node.oMatchesSelector then 'oMatchesSelector'
+    else if node.msMatchesSelector then 'msMatchesSelector'
+
+  if matchesMethod then getMethod(node, matchesMethod)
+  else (selector) ->
+    matchingNodes = getMethod(this, 'querySelector').call(this, selector)
```
Comment (jmsmyth) [inline modules/selection/main/index.coffee]: `this` isn't the right thing to be passing in here...

File: modules/selection/main/index.coffee (hunk 4 above)
Comment (jmsmyth) [inline modules/selection/main/index.coffee]: Using `.filter` will probably be less performant

File: modules/selection/main/index.coffee (hunk 1 above)
Comment (c-frater) [inline modules/selection/main/index.coffee]: Potentially, although it looks like it would have to still check for 'node.matches' first and then use hx.vendor if that doesn't exist

File: modules/selection/main/index.coffee (hunk 2 above)
Comment (c-frater) [inline modules/selection/main/index.coffee]: +1. It might also be worth double checking that this function works in the browser(s) that don't support the 'real' matches (If there are any left...)

File: modules/selection/main/index.coffee (hunk 3 above)
Comment (c-frater) [inline modules/selection/main/index.coffee]: The 'polyfill' was taken from the MDN which is why it's a bit odd:
```js
​function matches(elm, selector) {
    var matches = (elm.document || elm.ownerDocument).querySelectorAll(selector),
        i = matches.length;
    while (--i >= 0 && matches.item(i) !== elm) {}
    return i > -1;
}
```

Looking at it, the implementation is doing something different from the polyfill anyway (which is wrong and won't work)

File: modules/selection/main/index.coffee (hunk 4 above)
Comment (c-frater) [inline modules/selection/main/index.coffee]: +1 - I wanted to keep it efficient by using a for loop :)

File: modules/selection/main/index.coffee (hunk 5 above)
Comment (c-frater) [inline modules/selection/main/index.coffee]: Yeah, this 'polyfill' is wrong for multiple reasons....

File: modules/selection/main/index.coffee [hunk 6]
```diff
@@ -16,9 +16,34 @@ getMethod = (node, methodName) ->
...
 
+
+# Should only be called with Function.call(node, selector)
+matchPolyfill = (selector) ->
+  node = this
+  matchingNodes = (node.document or node.ownerDocument).querySelectorAll(selector)
+  [].slice.call(matchingNodes).indexOf(node) > -1
+
+getMatches = (node) ->
+  node.matches or hx.vendor(node, 'matchesSelector') or matchPolyfill
```
Comment (c-frater) [inline modules/selection/main/index.coffee]: Uses hx.vendor - This is assuming that we don't need to do the 'getMethod' style of call?
